riot-os-posix-lwip/client.c

```diff
@@ -0,0 +1,180 @@
+#include <string.h>
+
+/* socket includes */
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <unistd.h>
+#include "wolfssl/ssl.h"
+#include "wolfssl/certs_test.h"
+
+#define SERVER_PORT 11111
+
+
+int tls_client(int argc, char *argv[])
+{
+    int                sockfd;
+    char               buff[] = "Hello, TLS Server!\0";
+    char               *server_ip;
+    size_t             len;
+#ifdef LWIP_IPV6
+    struct sockaddr_in6 servAddr;
+#else
+    struct sockaddr_in servAddr;
+#endif
+
+    /* declare wolfSSL objects */
+    WOLFSSL_CTX* ctx;
+    WOLFSSL*     ssl;
+
+    puts("This is a TLS Client!");
+    if (argc != 2) {
+        fprintf(stderr, "Usage: %s IP_ADDR_SERVER\n", argv[0]);
+        return -1;
+    }
+    printf("Client is connecting to server at address %s port 11111...\n", argv[1]);
+    server_ip = argv[1];
+
+#ifdef LWIP_IPV6
+    /* Create a socket that uses an internet IPv6 address,
+     * Sets the socket to be stream based (TCP),
+     * 0 means choose the default protocol. */
+    if ((sockfd = socket(AF_INET6, SOCK_STREAM, 0)) == -1) {
+        fprintf(stderr, "ERROR: failed to create the socket\n");
+        return -1;
+    }
+
+    /* Initialize the server address struct with zeros */
+    memset(&servAddr, 0, sizeof(servAddr));
+
+    /* Fill in the server address */
+    servAddr.sin6_family       = AF_INET6;             /* using IPv6      */
+    servAddr.sin6_port         = htons(SERVER_PORT);  /* on SERVER_PORT */
+
+    /* Get the server IPv6 address from the compile-time string parameter */
+    if (inet_pton(AF_INET6, server_ip, &servAddr.sin6_addr.s6_addr) != 1) {
+        fprintf(stderr, "ERROR: invalid address\n");
+        return -1;
+    }
+#else
+    /* Create a socket that uses an internet IPv4 address,
+     * Sets the socket to be stream based (TCP),
+     * 0 means choose the default protocol. */
+    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) == -1) {
+        fprintf(stderr, "ERROR: failed to create the socket\n");
+        return -1;
+    }
+
+    /* Initialize the server address struct with zeros */
+    memset(&servAddr, 0, sizeof(servAddr));
+
+    /* Fill in the server address */
+    servAddr.sin_family       = AF_INET;             /* using IPv4      */
+    servAddr.sin_port         = htons(SERVER_PORT);  /* on SERVER_PORT */
+
+    /* Get the server IPv4 address from the compile-time string parameter */
+    if (inet_pton(AF_INET, server_ip, &servAddr.sin_addr.s_addr) != 1) {
+        fprintf(stderr, "ERROR: invalid address\n");
+        return -1;
+    }
+#endif
+    
+    /* Connect to the server */
+    if (connect(sockfd, (struct sockaddr*) &servAddr, sizeof(servAddr)) == -1) {
+        fprintf(stderr, "ERROR: failed to connect\n");
+        return -1;
+    }
+
+    /* Initialize wolfSSL */
+    wolfSSL_Init();
+
+    /* Create and initialize WOLFSSL_CTX */
+    if ((ctx = wolfSSL_CTX_new(wolfTLSv1_2_client_method())) == NULL) {
```

Comment:
Would TLS 1.3 work?

Reply:
TLS1.3 can be switched on by adding `USEMODULE+=wolfssl_tls13` to the makefile.

TLS1.3 in RIOT depends on `sock_tls`, I've submitted a PR to remove this forced dependency and make TLS1.3 available for POSIX/lwIP sockets.

https://github.com/RIOT-OS/RIOT/pull/16266